internal/ingress/controller/store/store.go

```diff
@@ -479,6 +479,18 @@ func New(checkOCSP bool,
 					if key == configmap {
 						store.setConfig(cm)
 					}
+
+					ings := store.listers.IngressAnnotation.List()
```

Comment:
While this is going to fix the issue, it seems hacky to me. I think the root of the issue is the way IP whitelisting is implemented at https://github.com/kubernetes/ingress-nginx/blob/fe9a5aec44242d32ca950d704623a94397522918/internal/ingress/annotations/ipwhitelist/main.go#L81

This is adding an extra cost that we can avoid by fixing root cause of the issue:

Reading https://github.com/kubernetes/ingress-nginx/blob/master/internal/ingress/annotations/ipwhitelist/main.go#L80 what it's doing is if `whitelist-source-range` annotation is not set then we use `whitelist-source-range` value from the configmap otherwise we use the one from the annotation. IMO that logic should not be part of the annotation parsing - annotation parsing should not be concerned about configmap. I suggest we move that logic into the template, in other words in the template we first check `location.Whitelist.CIDR` if it is not empty we then use that to generate `geo ...` config otherwise we use `$cfg.WhitelistSourceRange` to do so.

Reply:
I share that opinion, the improvement to the config detection is independent from the issue related to `whitelist-source-range`, this should be made clearer via a separate PR.

Reply:
@antoineco my comment was mainly about the fact that this fix is a hack, it is not fixing the root of the issue and unnecessarily introducing extra work on configmap updates.

Reply:
> While this is going to fix the issue, it seems hacky to me. I think the root of the issue is the way IP whitelisting is implemented at

This approach is used in all the annotation parsing step. Changing that in this PR is not going to happen.

> IMO that logic should not be part of the annotation parsing - annotation parsing should not be concerned about configmap

You are right here, this should not be part of the annotation parsing but right now we have no alternative. This is the most important reason why I want to move to CRDs and don't use annotations anymore, not only because is complex but it doesn't scale and it's impossible to add the semantics we need in the configuration options (like a list of IP addresses and not a comma-separated string that can contain anything)

> I suggest we move that logic into the template

No, that would make that impossible to understand, even now the templating step contains too much logic.